src/interpolation/details/ArborX_InterpDetailsPolynomialBasis.hpp

```diff
@@ -0,0 +1,152 @@
+/****************************************************************************
+ * Copyright (c) 2023 by the ArborX authors                                 *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_INTERP_DETAILS_POLYNOMIAL_BASIS_HPP
+#define ARBORX_INTERP_DETAILS_POLYNOMIAL_BASIS_HPP
+
+#include <Kokkos_Core.hpp>
+
+#include <type_traits>
+
+namespace ArborX::Interpolation
+{
+
+namespace Details
+{
+
+// Polynomial basis is computed in the same way as the following example
+// diagrams. Each cell is the product of the variable times everything on the
+// left cell and above.
+//
+// For example, the "degree 3 / variable y" has size 3 because it multiplies
+// every element of "degree 2 / variable y" and "degree 2 / variable x".
+//
+//   Quadratic |  3D  ||        Cubic      |  2D
+// ------------+----- || ------------------+-----
+//    degree   | var  ||       degree      | var
+// ---+---+----+      || ---+---+----+-----+
+//  0 | 1 |  2 |      ||  0 | 1 |  2 |  3  |
+// ---+---+----+      || ---+---+----+-----+
+// ---+---+----+      || ---+---+----+-----+
+//  1 |   |    |      ||  1 |   |    |     |
+// ---+---+----+----- || ---+---+----+-----+-----
+//    | x |    |  x   ||    | x |    |     |  x
+//    |   | xx |      ||    |   | xx |     |
+//    +---+----+----- ||    |   |    | xxx |
+//    | y |    |  y   ||    +---+----+-----+-----
+//    |   | xy |      ||    | y |    |     |  y
+//    |   | yy |      ||    |   | xy |     |
+//    +---+----+----- ||    |   | yy |     |
+//    | z |    |  z   ||    |   |    | xxy |
+//    |   | xz |      ||    |   |    | xyy |
+//    |   | yz |      ||    |   |    | yyy |
+//    |   | zz |      ||
+
+// The size of each cell is 1 if it is at degree 1 or at variable x
```

Comment:
It is the number of elements inside what I call "a cell", as referred just above. I may need to produce better diagrams/explanations as it may be a bit hard to visualize.

Reply:
To be exact, the cell at degree `deg` and dimension `dim` is the list of monomials of degree `deg` and of coordinates of dimensions `dim` and under.

I have renamed them to "slice" as it is closer to what it is, a slice of the evaluated polynomial basis.